crates/protocol/src/fee.rs

```diff
@@ -0,0 +1,483 @@
+    }));
+
+    rollup_data_gas_cost
+}
+
+/// Calculate the data gas for posting the transaction on L1.
+///
+/// In fjord, Calldata costs 16 gas per byte after compression.
+fn data_gas_fjord(input: &[u8]) -> U256 {
+    let estimated_size = tx_estimated_size_fjord(input);
+    return estimated_size
+        .saturating_mul(U256::from(NON_ZERO_BYTE_COST))
+        .wrapping_div(U256::from(1_000_000));
+}
+
+// Calculate the estimated compressed transaction size in bytes, scaled by 1e6.
+// This value is computed based on the following formula:
+// max(minTransactionSize, intercept + fastlzCoef*fastlzSize)
+fn tx_estimated_size_fjord(input: &[u8]) -> U256 {
+    let fastlz_size = U256::from(flz_compress_len(input));
+
+    fastlz_size
+        .saturating_mul(U256::from(836_500))
+        .saturating_sub(U256::from(42_585_600))
+        .max(U256::from(100_000_000))
+}
+
+/// Calculate the gas cost of a transaction based on L1 block data posted on L2 post-bedrock.
+pub fn calculate_tx_l1_cost_bedrock(
+    input: &[u8],
+    l1_fee_overhead: U256,
+    base_fee: U256,
+    l1_fee_scalar: U256,
+    is_before_regolith: bool,
+) -> U256 {
+    if input.is_empty() || input.first() == Some(&0x7F) {
+        return U256::ZERO;
+    }
+
+    let rollup_data_gas_cost = data_gas_bedrock(input, is_before_regolith);
+    rollup_data_gas_cost
+        .saturating_add(l1_fee_overhead)
+        .saturating_mul(base_fee)
+        .saturating_mul(l1_fee_scalar)
+        .wrapping_div(U256::from(1_000_000))
+}
+
+/// Calculate the gas cost of a transaction based on L1 block data posted on L2 post-regolith.
+pub fn calculate_tx_l1_cost_regolith(
+    input: &[u8],
+    l1_fee_overhead: U256,
+    base_fee: U256,
+    l1_fee_scalar: U256,
+) -> U256 {
+    if input.is_empty() || input.first() == Some(&0x7F) {
+        return U256::ZERO;
+    }
+
+    let rollup_data_gas_cost = data_gas_regolith(input);
+    rollup_data_gas_cost
+        .saturating_add(l1_fee_overhead)
+        .saturating_mul(base_fee)
+        .saturating_mul(l1_fee_scalar)
+        .wrapping_div(U256::from(1_000_000))
+}
+
+/// Calculate the gas cost of a transaction based on L1 block data posted on L2, post-Ecotone.
+///
+/// L1 cost function:
+/// `(calldataGas/16)*(l1BaseFee*16*l1BaseFeeScalar + l1BlobBaseFee*l1BlobBaseFeeScalar)/1e6`
+///
+/// We divide "calldataGas" by 16 to change from units of calldata gas to "estimated # of bytes when compressed".
+/// Known as "compressedTxSize" in the spec.
+///
+/// Function is actually computed as follows for better precision under integer arithmetic:
+/// `calldataGas*(l1BaseFee*16*l1BaseFeeScalar + l1BlobBaseFee*l1BlobBaseFeeScalar)/16e6`
+pub fn calculate_tx_l1_cost_ecotone(
+    input: &[u8],
+    l1_fee_overhead: U256,
+    l1_fee_scalar: U256,
+    base_fee: U256,
+    base_fee_scalar: U256,
+    blob_base_fee: U256,
+    blob_base_fee_scalar: U256,
+    empty_scalars: bool,
```

Comment:
I don't think we need to take this into this function; the consumer can decide to use the Bedrock fee function at a higher level.